### ISSUE TITLE
Conflicts getting ahead of queued pushes?

### ISSUE THREAD

**Issue by michielbdejong** [remotestorage/remotestorage.js#804]
[this starter-kit bug](https://github.com/remotestorage/starter-kit/issues/21) looks like a problem with the library. It is easy to replicate, @silverbucket and I have both seen it happen.

It looks like maybe there are conflicts coming in while outgoing pushes are still being queued. Maybe they're queued in https://github.com/remotestorage/remotestorage.js/blob/master/src/cachinglayer.js#L306-L315, maybe in the commit-cache, maybe just in sync tasks. Does make me think that maybe we have too many queues! ;) Should nonetheless be easy to fix, if you just reproduce it and trace what happens.

**Comment by michielbdejong** [comment]
@silverbucket You added the ready label but then @skddc removed it... This is quite serious though, if it even makes the starter-kit hello world example fail. We should probably find time to fix this. Do you want to pair on this?

**Comment by skddc** [comment]
I removed it, because we don't use that label. It was automatically added by a tool, that only @silverbucket uses personally.

**Comment by silverbucket** [comment]
yeah, sorry about that - i've since disabled that feature (was doing it automatically)

**Comment by karlb** [comment]
Any updates on this? I think I'm still experiencing this problem with v2.0.0-beta.6 (but also earlier versions).

**Comment by raucao** [comment]
Wow, this is one is ancient!

@karlb Could you recap the problem real quick, and ideally provide an example to reproduce it? The starter kit that the bug has initially been created for has been deprecated for a very long time.

**Comment by karlb** [comment]
I don't have a minimal example, but when I type quickly at https://static.karl.berlin/doagain/ while being signed in to my 5apps remote storage, I get a conflict despite my browser being the only device sending writes. This sounded similar to this issue, but without further research it is just a guess.

**Comment by raucao** [comment]
Oof, that app is doing immediate writes to the same document on every keystroke with zero [debouncing](https://dev.to/jeremiahjacinth13/what-is-debouncing-1akk). That's definitely never a good idea, aside from potential race conditions in the library!

However, I think the bug's explanation is that outgoing pushes/PUTs to the server are queued up for successive sync immediately. And when there's a push still outgoing that hasn't updated the current revision (ETAG) yet, then the already queued next push would use the outdated ETAG and thus fail. (Just recapping the issue to be solved).

**Comment by karlb** [comment]
> Oof, that app is doing immediate writes to the same document on every keystroke with zero [debouncing](https://dev.to/jeremiahjacinth13/what-is-debouncing-1akk). That's definitely never a good idea, aside from potential race conditions in the library!

You are totally right of course and I also noticed while looking at the problem. I just wanted to understand the conflict problem before, so that I don't just hide the problem. Now that I improved my conflict handling and the bug is at least tracked properly, I'll go and add the debouncing.

Thanks for looking into it and your (and everyone else's) work remotestorage!

**Comment by DougReeder** [comment]
As debouncing is required, that ought to be the responsibility of the library, rather than every single app that uses it. I'd be happy to work on that once this issue is resolved.

**Comment by raucao** [comment]
I just had a look at this, because I improved the library's conflict events (also live on https://myfavoritedrinks.remotestorage.io now). However, i wasn't able to reproduce the bug with https://static.karl.berlin/doagain/.

I did see that this particular app is storing all tasks in a single JSON document, which is definitely not recommended. For minimizing sync issues, it's always best to store different objects in different documents, regardless of any conflict handling in place. The best conflicts are simply no conflicts

Since pushes are happening very quickly under normal circumstances, and no revisions are changed remotely while being offline and queuing up more pushes, the only scenario I can imagine where this is actually an issue is exactly the example from above: no debouncing while attempting updates on every keystroke of user input, which was also the case in the original starter kit issue.

So I made My Favorite Drinks update documents on every keystroke. However, even though I did see it happen once, it's still almost impossible for me to reproduce the issue, meaning that the queuing generally works pretty well. Here's a sceencast of me typing and observing PUTs:

https://github.com/user-attachments/assets/c5cae5ab-b141-4db7-8839-49d8b7a47a77

(This is on the latest #1331 branch, so maybe something was already fixed or improved there. I haven't tested with anything else yet.)

I think debouncing should generally solve this edge case well enough, especially in combination with using separate documents for separate data objects. However, if it's actually a problem in someone's app, and there are no other ways of mitigating it, then the conflict event can be used exactly for what it's for: reading the conflicting values from it and applying local conflict resolution.

Considering the above, I'll close this issue for now. But please feel free to add new comments and feedback, so we can also re-open it if necessary or desired.
